docs/formal-review.md

```diff
@@ -0,0 +1,146 @@
+
+If you changed the method, review use sites of this method.
+
+# New code errors
+
+## Side effects
+
+### Data invariants
+
+Objects should be constructed with invariants satisfied. If possible, there should be no invariant exceptions for “uninitialized” objects.
+
+### Create do not reuse
+
+If it is possible, I should construct new objects instead of reusing existing ones.
+
+### Separate inputs and outputs
+
+If possible, signify outputs in the result and inputs in arguments. Reference arguments should be avoided as much as possible, especially in new code.
+
+### Type (conversion) errors
+
+C++ has a standardized type constraint solution engine and it is possible to have a program logic error with type conversions and assignments. Myself, I prefer not to implement assignment and conversion operators (and operators at all, if that’s possible) but other’s code can contain these.
+
+
+### Change of structure of one of the types
+
+Usually, the deletion or change of field is flagged in compiler errors. The addition of fields is not tracked at all, thus, this is the source of errors in assignment operators and other structure-copying mechanisms.
+
+Tracking can be implemented with the change of class/type’s name, then reverting the change back when everything is done.
+
+# Structure of code - design
+
+## Execution should avoid interpretation
+
+Execution of some action should avoid conditions on the data as much as possible.
+
+This warrants a concrete example.
+
+Instead of
+
+```
+if (condition on the fields)
+        { auto p = createThisObj(...); p-> bubu(); }
+else { auto p = createThatObj(...); p->gaga() }
+```
+
+We should create object with either executor (or both) and execute them conditionally:
+
+```
+If (thisObjBubuExec) { thisObjBubuExec->bubu(); }
+If (thatObjGagaExec) { thatObjGagaExec->gaga(); }
+```
+
+First variant pushes optimizations and rewrites of execution down to the execution phase, instead of pulling them up into optimization and rewrite phase.
+
+## Execution must be stupid. It is the optimization that must be smart.
+
+### Templates execute, not decide
+
+It is hard to inspect control decisions inside templated code. For example, not all template parametrizations may have serialization to streams implemented. Because of that, templated code should have as few decisions as possible.
+
+It should be preferred to call templates with all decisions made elsewhere and for templates to execute only simple decisions like loops.
+
+## Inheritance should be shallow
+
+Liskov substitution principle usually can be realized several ways. Classical example is the Square and Rectangle inheritance - which one should inherit the other? One can argue both ways, that Rectangle adds dimension and should inherit Square and that Square is a constrained Rectangle and thus should be inherited from.
```

Comment:
Clean architecture, huh?

Reply:
Class hierarchy according with Martin enables dependency reversal so it is for the better. You define an iface that tends to change less often and here you go the dependency inversion. And this is C++ in the end. I would say this is too generic to get into this doc.